Add default implementation of the "About" command

saroad2 opened the pull request:
Added a default implementation of the "About" command. This command should work absolutely fine in any of the supported platforms, as long as `info_dialog` is implemented correctly.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] All new features have been tested
- [X] All new features have been documented
- [X] I have read the **CONTRIBUTING.md** file
- [X] I will abide by the code of conduct

freakboy3742 commented on src/cocoa/toga_cocoa/app.py:
```diff
@@ -123,7 +123,6 @@ def create(self):
         formal_name = self.interface.formal_name
 
         self.interface.commands.add(
-            toga.Command(None, 'About ' + formal_name, group=toga.Group.APP),
```
I understand the intention here, but I don't think this will work. In an idiomatic macOS app, `About` appears in the app menu, not the help menu, and has the name "About MyApp".

I suspect that the menu *construction* will need to remain an impl-layer thing; although the actual commands can be made interface layer.

freakboy3742 commented on src/core/toga/app.py:
```diff
@@ -227,6 +230,9 @@ def __init__(
         self.on_exit = on_exit
 
+        if default_about_command:
+            self.commands.add(AboutCommand(self))
```
This hard-codes the about dialog that will be used. While the default will definitely be sufficient for many purposes, there will always be a call for customisation.

I'd suggest that `about()` should be a method on the app interface, with a default implementation that creates a default about dialog. While we're in the area, we should probably also factor out the other commands into similar utility methods (preferences, homepage etc) - even if they're no-ops for now.

Making `about()` an app-level method that can be overwritten also means we can remove the `default_about_command` option - a Toga will get a reasonable default, and if you want to override, that's easy enough to acheive with an override.

freakboy3742 reviewed (changes requested):
This is a good start; but there's some details that need cleaning up - mostly around ensuring that the commands remain platform-idiomatic.

Longer term, I'd like to look at replacing the use of `info_dialog`, too. Using a dialog is a fine way to get *something* working, but "real app" about dialogs don't present as modal info dialogs on macOS.

saroad2 commented:
At the moment, keep using `info_dialog` as I initially implemented. If you have an idea of how to do it better in Mac or Windows, please let me know.

Added the implementation in all platforms, along with all the other default commands (preferences, home_page and quit).

Once you approve this PR I will implement a default `QuitCommand` similar to the `AboutCommand` I already implemented.

freakboy3742 reviewed (changes requested):
I think we've had a miscommunication here. I don't think the *command* is what needs to be customizable; it's the action performed *by* the command that needs to be customized. 

The keyboard shortcuts, menu positions etc of the "required" menu items essentially aren't negotiable for end-developers (at least, they aren't negotiable if you want to remain platform-HIG compliant). The "quit" command will always have the same name, the same keyboard shortcut, and the same position in the overall menu structure. However, the *behavior* when you quit might need to change e.g., performing "on-exit" behavior, or changing the appearance of the "about" dialog. 

So - rather than have extra arguments to `App`, I'd suggest that `preferences()`, `about()`, `quit()` etc (possibly `on_about()`...?) be methods on the interface-level `App`, that can be overwritten by subclasses if they need to do so. The declaration of the commands themselves can essentially remain as they are in master, but invoking the interface-level entry points.

(As an aside: #980 will have some interesting interactions when you get to the "quit" menu item)

saroad2 commented:
I'm not sure I fully agree with this approach.

I'll give you my own example: In *Eddington* I want to change the **Preference** command to be in a different section in the **File** group. At the moment it is in the first section. In my opinion, we should support a way to override it.

freakboy3742 commented:
So, if you did that, you'd be in violation of the [macOS HIG](https://developer.apple.com/design/human-interface-guidelines/macos/app-architecture/preferences/). By that document, App preferences should be listed in the App menu. Windows has [similar guidance](
https://docs.microsoft.com/en-us/windows/uwp/design/app-settings/guidelines-for-app-settings).

Not that this only applies to the *application* preferences. This doesn't preclude having a *separate* preferences menu for some specific aspect of your app. For example, Xcode has a *lot* of preferences/configuration dialogs for different pieces of the GUI - but it also has an "App Preferences" for setting up your developer credentials, editor preferences, and so on.

One of the design goals I've got for Toga is to make Toga apps adhere to the HIG by default, and prevent divergence from the HIG wherever possible. Making sure the core menu items are all in the right place is a key piece of this.

saroad2 commented:
I see.

No problem. I'll fix that.

saroad2 commented:
Since it would be a lot of changes, closing this PR and opening a new one.
